### PR TITLE
Add symbol search

### PR DESCRIPTION
**Pull request by artmoskvin**
- new handler `/project/.../search?type=symbol&query=...&limit=10`
- powered by LSP servers
- filter out low level symbols like fields and variables from search results; currently hardcoded but can be exposed as a query parameter later

### CODE REVIEW

File: pkg/project/manager_test.go [hunk 1]
```diff
@@ -186,3 +190,72 @@ func TestManagerImpl_CreateTask_ExecError(t *testing.T) {
 		t.Errorf("Expected error, got nil")
 	}
 }
+
+func TestManagerImpl_GetWorkspaceSymbols(t *testing.T) {
```
Comment (artmoskvin) [inline pkg/project/manager_test.go:194]:
```suggestion
func TestManagerImpl_SearchSymbols(t *testing.T) {
```

File: pkg/project/manager.go [hunk 2]
```diff
@@ -508,6 +509,25 @@ func (pm ManagerImpl) UpdateLines(ctx context.Context, projectId, path string, l
 	return file, nil
 }
 
+func (pm ManagerImpl) SearchSymbols(ctx context.Context, projectId model.ProjectId, query string) ([]lsp.SymbolInfo, error) {
```
Comment (artmoskvin) [inline pkg/project/manager.go:512]: check context cancellation

File: pkg/handlers/search_symbols.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+type SearchSymbolsHandler struct {
+	pm         project.Manager
+	lspService lsp.Service
+}
+
+func NewSearchSymbolsHandler(pm project.Manager, lspService lsp.Service) SearchSymbolsHandler {
+	return SearchSymbolsHandler{pm: pm, lspService: lspService}
+}
+
+func (h SearchSymbolsHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```
Comment (artmoskvin) [inline pkg/handlers/search_symbols.go:27]: test

File: pkg/handlers/search_symbols.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+const MinLimit = 1
+const MaxLimit = 100
+const DefaultLimit = 10
+
+type SearchSymbolsHandler struct {
+	pm project.Manager
+}
+
+func NewSearchSymbolsHandler(pm project.Manager) SearchSymbolsHandler {
```
Comment (aleh-null) [inline pkg/handlers/search_symbols.go:21]: Let's add some options to the initialiser like `limit := DefaultLimit`, I prefer to have those upfront on initialisation also can make the service more configurable once we get to the config. [Functional Options Pattern](https://matheuspolitano.medium.com/unlocking-the-power-of-functional-options-pattern-in-go-087478f57be9) is nice, I typically use it but in this case struct in `SearchSymbolsHandler` will do.

File: pkg/handlers/search_symbols.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+	query := r.URL.Query().Get("query")
+	if query == "" {
+		http.Error(w, "query not specified", http.StatusBadRequest)
+		return
+	}
+
+	limit := DefaultLimit
+
+	if limitParam := r.URL.Query().Get("limit"); limitParam != "" {
```
Comment (aleh-null) [inline pkg/handlers/search_symbols.go:40]: I think we should use `Has()`, try to parse and through an error.  https://go.dev/play/p/Ech5tabjRP0

File: pkg/handlers/search_symbols.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+		limit, err = strconv.Atoi(limitParam)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("invalid limit %s: %s", limitParam, err), http.StatusBadRequest)
+			return
+		}
+
+		if limit < MinLimit || limit > MaxLimit {
+			http.Error(w, "limit must be between 1 and 100", http.StatusBadRequest)
```
Comment (aleh-null) [inline pkg/handlers/search_symbols.go:48]:
```suggestion
			http.Error(w, fmt.Sprintf("limit must be between %v and %v",  MinLimit, MaxLimit), http.StatusBadRequest)
```

File: cmd/run.go [hunk 7]
```diff
@@ -103,7 +104,7 @@ var runCmd = &cobra.Command{
 		languageDetector := lsp.NewFileExtensionBasedLanguageDetector()
 		diagnosticsStore := lsp.NewDiagnosticsStore()
 		clientPool := lsp.NewClientPool()
-		lspService := lsp.NewService(languageDetector, lspServerExecutables, diagnosticsStore, clientPool)
+		lspService := lsp.NewService(languageDetector, lspServerExecutables, diagnosticsStore, clientPool, []protocol.SymbolKind{protocol.SymbolKindField, protocol.SymbolKindVariable})
```
Comment (aleh-null) [inline cmd/run.go:107]: I think those options should be passed to `SearchSymbols()` by the caller func. This should add some flexibility to the codebase.

File: pkg/handlers/search_symbols.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+			return
+		}
+	}
+
+	symbols, err := h.pm.SearchSymbols(r.Context(), projectID, query)
```
Comment (aleh-null) [inline pkg/handlers/search_symbols.go:53]: As mentioned previously I would pass ignore symbols as options here.

**Review by aleh-null — COMMENTED** [review]
LGTM. I have some nits on the options. Let's discuss.

File: pkg/handlers/search_symbols.go (hunk 4 above)
Comment (artmoskvin) [reply pkg/handlers/search_symbols.go]: Done

File: pkg/handlers/search_symbols.go (hunk 5 above)
Comment (artmoskvin) [reply pkg/handlers/search_symbols.go]: Done

File: pkg/handlers/search_symbols.go (hunk 6 above)
Comment (artmoskvin) [reply pkg/handlers/search_symbols.go]: Done

File: cmd/run.go (hunk 7 above)
Comment (artmoskvin) [reply cmd/run.go]: Done

File: pkg/handlers/search_symbols.go (hunk 8 above)
Comment (artmoskvin) [reply pkg/handlers/search_symbols.go]: Done

**Review by aleh-null — APPROVED** [review]

**Pull request merged**